### ISSUE TITLE
Learning rate adjustment in async-distributed training

### ISSUE THREAD

**Issue by skydoorkai** [sql-machine-learning/elasticdl#89]
Assume a model converges under 1 worker with learning rate _lr_. When switches to async-distributed training with _n_ workers, the same learning rate  _lr_ might not work. This is because that _n_  workers may get the same version of a model, and each computes a gradient and sends to ps. Ps will apply the _n_ gradients one-by-one, this is equivalent of using _n*lr_ as learning rate in sync-distributed training.

Dividing the learning rate by the number of workers is a way to ease this issue. Below is the test results of an epoch training with mnist data with different worker numbers (n=1, 2, 4, 8) and with/without learning rate scale (1/n).  For example, 'w2_scale' means 2 workers with learning rate scale. 'w4_no_scale' means 4 workers without learning rate scale.

learning_rate = 0.01
batch_size = 64

![figure_1](https://user-images.githubusercontent.com/16055370/50157895-8f96e580-030d-11e9-92f3-9a265740ed1e.png)

We can see with n=2,4,8, training accuracy remains 0.1 without learning rate scale.

With learning rate scale, n=2,4 converges, but n= 8 is not. This is another issue with async-distributed training as the number of worker increases. Some gradients are too old to the model version in ps, which will prevent the training accuracy improvement. One way to address this issue is to adjust the learning rate further for the old gradients or just drop them.

Below is the test result of staleness-aware learning rate adjustment.
We define 
_staleness = difference with the worker gradient version and ps version_
In 1-worker scenario, _staleness = 1_  
In 2-worker scenario, if worker 0 and 1 use the same version of the model, compute their gradients, and push the gradients to ps. The 1st applied gradient has _staleness = 1_, the second applied gradient has _staleness = 2_.
Learning rate _lr_ is adjusted by _staleness_:
_lr_used = lr / staleness_

![staleness_aware](https://user-images.githubusercontent.com/16055370/50205462-472a0700-03a3-11e9-9973-185d39cb7d06.png)

Model converges with worker number _n=1, 2, 4, 8_, but does not with _n=16_.

Thus, we can see that async-distributed training might not suitable for cnn model with large number of workers.  Probably that is why almost all distributed training for cnn model use synchronized strategy.
